versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -757,7 +761,21 @@ protected CommitLogEntry buildIndividualCommit(
             Collections.emptyList());
 
     if (keyListDistance >= config.getKeyListDistance()) {
-      entry = buildKeyList(ctx, entry, newKeyLists);
+      Hash startHash;
+      if (unwrittenEntriesForKeyList.isEmpty()) {
+        // query all commits from backend.
+        startHash = entry.getParents().get(0);
```

Comment:
the comment says query all commits but we do `get(0);`. Shouldn't it iterate over it as well?

Reply:
if the startHash points to head commit (aka `get(0)`), we iterate all the commits from startHash inside `buildKeyList -> keysForCommitEntry`